File: spkmeans.c
#include "spkmeans.h"
#include <string.h>
int rows, cols;


static void print_matrix(double** matrix){
    int i, j;
    for (i = 0; i < rows; i++) {
        for (j = 0; j < rows-1; j++)
            printf("%.4f%c", matrix[i][j], ',');
        printf("%.4f%s", matrix[i][j], "\n");
    }
}

static double** get_matrix(char* file_name){
    int i, j;
    double d;
    double** matrix;
    char c = 0;
    FILE *file_input;
    cols = 0, rows = 1;

    file_input = fopen(file_name, "r");
    if (file_input == NULL)
        return NULL;

    while (c != '\n') {
        fscanf(file_input, "%*f%c", &c);
        cols++;
    }
    while (fscanf(file_input, "%*s\n") != EOF)
        rows++;
    rewind(file_input);
    matrix = Array_2D(rows, cols, 1);
    if (matrix == NULL)
        return NULL;

    for (i = 0; i < rows; i++) {
        for (j = 0; j < cols; j++) {
            fscanf(file_input, "%lf%*c", &d);
            matrix[i][j] = d;
        }
    }
    fclose(file_input);
    return matrix;
}

int main(int argc, char* argv[]) {
    char goal[6];
    double **matrix, **W;

    if (argc != 3){
        printf("Invalid Input!");
        return 1;
    }
    sscanf(argv[1], "%s", goal);
    matrix = get_matrix(argv[2]);
    if (matrix == NULL) {
        printf("Invalid Input!");
        return 1;
    }

    if (strcmp(goal, "wam") == 0)
        matrix = wam(matrix, rows, cols);
    else if (strcmp(goal, "ddg") == 0) {
        W = wam(matrix, rows, cols);
        matrix = ddg(W, rows);
        free_Mem(W, rows);
    }
    else if (strcmp(goal, "lnorm") == 0)
        matrix = L_norm(matrix, rows, cols);
    else if (strcmp(goal, "jacobi") == 0){
        matrix = jacobi(matrix, rows);
        free_Mem(eigen_mat, rows);
        }
    else{
        printf("Invalid Input!");
        return 1;
    }

    if (matrix == NULL){
        printf("An Error Has Occurred");
        return 1;
    }

    print_matrix(matrix);
    free_Mem(matrix, rows);
    return 0;
}

File: spkmeans.py
import sys
import numpy as np
import numpy.random
import spkmeansmodule as kp
np.random.seed(0)


def euclidean_dist(vector1, vector2):
    dist = (vector1 - vector2) ** 2
    return np.sqrt(sum(dist))


def print_resault(resault, index, n):
    if index is not None:
        print(*index, sep=",")
    resault = ['%.4f' % resault[i] for i in range(len(resault))]
    resault = [resault[i:i + n] for i in range(0, len(resault), n)]
    for i in range(len(resault)):
        print(*(resault[i]), sep=",")


def kmeans_pp(vectors, k):
    index = []
    clusters = np.zeros((k, len(vectors[0])), dtype="float")
    first_cluster_index = np.random.choice(len(vectors))
    index.append(first_cluster_index)
    clusters[0] = np.copy(vectors[first_cluster_index])
    distances = np.full(len(vectors), sys.float_info.max)
    i = 1
    while i < k:
        for t in range(len(vectors)):  # compute distances from all vectors to the closest cluster
            for j in range(i):
                dist = euclidean_dist(vectors[t], clusters[j])
                if dist < distances[t]:
                    distances[t] = dist
        probabilities = distances / sum(distances)  # compute probabilities
        chosen_index = int(np.random.choice(len(vectors), 1, True, probabilities))
        clusters[i] = np.copy(vectors[chosen_index])
        index.append(chosen_index)
        i += 1
    return clusters, index


def k_legal(k):
    if k - int(k) > 0 or k < 0:
        print("Invalid Input!")
        return False
    return True


def execute_by_goal(vectors, goal):
    goal_to_function = {
        "wam": kp.wam,
        "ddg": kp.ddg,
        "lnorm": kp.lnorm,
        "jacobi": kp.jacobi,
    }
    return goal_to_function[goal](len(vectors), len(vectors[0]), vectors.flatten().tolist())


def create_vectors_from_file(file):
    try:
        vectors = np.genfromtxt(file, delimiter=",")
        return vectors
    except IOError:
        print("Invalid Input!")
        return None


def goal_is_spk(vectors, k):
    n = len(vectors[0])
    if not k_legal(k):
        return 1
    k = int(k)
    if k == 0:
        res = kp.spk(len(vectors), len(vectors[0]), vectors.flatten().tolist())
        if res is None:
            return None, None, None
        k = res[0]
        vectors = np.array(res[1]).reshape(int(len(res[1]) / k), k)
        n = int(k)

    init_clusters, chosen_index = kmeans_pp(vectors, k)
    clusters = kp.fit(k, len(vectors), len(vectors[0]), vectors.flatten().tolist(), init_clusters.flatten().tolist())
    return clusters, chosen_index, n


def main():
    cmd_input = sys.argv
    k = float(cmd_input[1])
    goal = cmd_input[2]
    file_name = cmd_input[3]

    vectors = create_vectors_from_file(file_name)
    if vectors is None:
        return 1

    chosen_index = None
    n = len(vectors)

    if goal == "spk":
        resualt, chosen_index, n = goal_is_spk(vectors, k)
    else:
        resualt = execute_by_goal(vectors, goal)
    if resualt is None:
        print("An Error Has Occurred")
        return 1

    print_resault(resualt, chosen_index, n)
    return 0


if __name__ == '__main__':
    main()
